L83-core-cpp-tls-key-export-feature.md

```diff
@@ -0,0 +1,202 @@
+  * `std::string tls_key_log_file_path`: Full path where the TLS keys should be exported to.
+  * `grpc_tls_key_log_format  tls_key_log_format`: Format of the exported keys.
+
+  The `grpc::experimental::TlsKeyLoggerConfig` class has the following signature:
+  ```
+  class TlsKeyLoggerConfig {
+    private:
+      // The full path at which the TLS keys would be exported.
+      std::string tls_key_log_file_path;
+
+      // The format in which the TLS keys would be exported at this path.
+      grpc_tls_key_log_format tls_key_log_format;
+    public:
+      // Constructor
+      explicit TlsKeyLoggerConfig();
+
+      // Copy ctor
+      TlsKeyLoggerConfig(const TlsKeyLoggerConfig& copy)
+
+      // Sets the tls key log file path.
+      void set_tls_key_log_file_path(std::string key_log_file_path);
+
+      // Sets the tls key logging format. Currently only NSS format is supported.
+      void set_tls_key_log_format(grpc_tls_key_log_format tls_key_log_format);
+
+      // Returns the set tls key log file path.
+      std::string get_tls_key_log_file_path();
+
+      // Returns the set tls key log file format.
+      grpc_tls_key_log_format get_tls_key_log_format();
+  };
+  ```
+
+* A `set_tls_key_log_config(grpc::experimental::TlsKeyLoggerConfig cfg)` method is added to the existing `grpc::experimental::TlsCredentialsOptions` class to set the desired TLS session key export configuration on the client/server side.
+
+These changes allow specifying TLS Key export configuration on a per-channel level on the client side and on a per-port level on the server side. If the configuration is not explicitly set, TLS key logging is disabled. This config  struct can be extended in the future to support advanced features such as key logging based on source ip filtering etc.
+
+For extra protection, these configurations only take effect when the `GRPC_TLS_KEY_LOGGING_ENABLED` environment variable is set to true.
+
+## C-Core API Additions
+
+The following additions are made to the gRPC C-Core API to enable TLS session key logging. The API additions are described below with inline comments.
+
+```
+/** --- TLS Key logging. ---
+ * Experimental API to control tls key logging. Tls key logging is expected
+ * to be used only for debugging purposes and never in production. Tls
+ * key logging is only enabled when:
+ * 1. At least one grpc_tls_credentials_options object is assigned a tls
+ *  key logging config using the API specified below.
+ * 2. The GRPC_TLS_KEY_LOGGING_ENABLED environment variable is set to true.
+ */
+
+/**
+ * An opaque type pointing to a Tls Key logger helper object which provides
+ * methods to log TLS session keys to files.
+ */
+typedef struct gprc_tls_key_logger grpc_tls_key_logger;
+
+/**
+ * An opaque type pointing to a Tls key logging configuration.
+ */
+typedef struct grpc_tls_key_log_config grpc_tls_key_log_config;
```

Comment:
Since `grpc_tls_key_log_config` is an opaque type, how can the users specify different configurations through it?
I thought we were going to have something like this:
```
typedef struct tls_key_log_format {
  ...
}

typedef struct grpc_tls_key_log_config {
  const char* log_file_path;
  tls_key_log_format format;
}
void grpc_tls_credentials_options_set_tls_key_log_config(
    grpc_tls_credentials_options* options, grpc_tls_key_log_config* config);
```
Then we don't need to expose `gprc_tls_key_logger` at all.